indra/newview/llagentiomodule.cpp

```diff
@@ -0,0 +1,279 @@
+{
+    //Camera and target position are expected as
+    //world-space coordinates in the viewer-standard
+    //coordinate frame.
+    if (data.has("camera") && data.has("target"))
+    {
+        LLUUID target_id = LLUUID::null;
+        
+        if (data.has("target_id"))
+        {
+            target_id = data["target_id"].asUUID();
+        }
+        
+        gAgentCamera.setCameraPosAndFocusGlobal(
+            ll_vector3d_from_sd ( data["camera"] ),
+            ll_vector3d_from_sd ( data["target"] ),
+            target_id);
+    }
+}
+
+void LLAgentIOModule::sendAgentOrientation()
+{
+    /*Find the agent's position and rotation in world and send it out*/
+    LLVector3 pos = LLVector3(gAgent.getPositionGlobal());
+    LLQuaternion rot = gAgent.getFrameAgent().getQuaternion();
+    
+    LLSD dat = LLSD::emptyMap();
+    dat["position"] = ll_sd_from_vector3(pos);
+    dat["rotation"] = ll_sd_from_quaternion(rot);
+
+    sendCommand("agent_orientation", dat);
+}
+
+void LLAgentIOModule::sendCameraOrientation()
+{
+    /*Send camera position and facing relative the agent's
+     position and facing.*/
+    LLVector3 camera_pos= gAgentCamera.getCurrentCameraOffset();
+    LLVector3 target_pos = LLVector3(gAgentCamera.getCurrentFocusOffset());
+    
+    LLSD dat = LLSD::emptyMap();
+    dat["camera"] = ll_sd_from_vector3(camera_pos);
+    dat["target"] = ll_sd_from_vector3(target_pos);
+
+    sendCommand("viewer_camera", dat);
+}
+
+void LLAgentIOModule::sendLookAt()
+{
+    //Send what the agent is looking at to the leap module.
+    //If the agent isn't looking at anything, sends empty map
+    //Otherwise, sends the direction relative the avatar's
+    //facing and the distance to the look target.
+    LLPhysicsMotionController::ptr_t motion(std::static_pointer_cast<LLPhysicsMotionController>(gAgentAvatarp->findMotion(ANIM_AGENT_PHYSICS_MOTION)));
+    
+    if (!motion)
+    {
+        LL_WARNS("AgentIO") << "Agent has no physics motion" << LL_ENDL;
+        return;
+    }
+        
+    LLVector3* targetPos = (LLVector3*)motion->getCharacter()->getAnimationData("LookAtPoint");
+
+    LLSD dat = LLSD::emptyMap();
+
+    if (targetPos)
+    {
+        LLVector3 headLookAt = *targetPos;
+        
+        dat["direction"] = ll_sd_from_vector3(headLookAt);
+        dat["distance"] = (F32)headLookAt.normVec();
+    }
+
+    sendCommand("look_at", dat);
+}
+
+LLAgentIOModule::LLAgentIOModule(const LL::LazyEventAPIParams& params) :
+    LLEventAPI(params),
+    mLeapModule()
+{
+    //NOTE: This debug line ensures the module gets linked.
+    LL_DEBUGS("LeapAgentIO") << "Initialized." << LL_ENDL;
+    AgentIOModule = this;
+}
+
+
+void LLAgentIOModule::setLeapModule(std::weak_ptr<LLLeap> mod, const std::string & module_name)
+{
+    mLeapModule = mod;
+    mModuleName = module_name;
+};
+
+
+LLAgentIOModule::module_ptr_t LLAgentIOModule::getLeapModule() const
+{
+    // lock it
+    return mLeapModule.lock();
+}
+
```

Comment:
Instead of a `sendCommand()` method, how about instantiating [a `Response` object](https://github.com/secondlife/viewer/blob/aura/DRTVWR-558-dbljump/indra/llcommon/lleventapi.h#L72-L153) in each method that wants to send a response? Or -- the heart of `Response` -- just call [sendReply()](https://github.com/secondlife/viewer/blob/aura/DRTVWR-558-dbljump/indra/llcommon/llevents.h#L771-L783).

In my mind, an `LLEventPump` alias such as `"agentio.controller"` is more for sending unsolicited data to a plugin than it is for replying to an explicit request.